torchreid/engine/image/am_softmax.py

```diff
@@ -377,4 +412,175 @@ def _apply_batch_transform(self, imgs, obj_ids):
+                optimizer = torch.optim.SGD(self.models[name].parameters(), lr=min_lr, weight_decay=wd)
+            else:
+                optimizer = self.optims[name]
+            if epochs_warmup != 0:
+                state_cacher = StateCacher(in_memory=True, cache_dir=None)
+                state_cacher.store("model", self.models[name].module.cpu().state_dict())
+                state_cacher.store("optimizer", self.optims[name].state_dict())
+                self.models[name].module.to(model_device)
+                print("Warmup the model's weights for {} epochs".format(epochs_warmup))
+                self.run(max_epoch=epochs_warmup, lr_finder=True)
+                print("Finished warmuping the model. Continue to find learning rate:")
+            # run lr finder
+            lr_finder = LRFinder(self.models[name], optimizer, criterion, device=model_device)
+            lr_finder.range_test(self.train_loader, start_lr=min_lr, end_lr=max_lr, smooth_f=0.01, num_iter=self.num_batches, step_mode='exp')
+            ax, optim_lr = lr_finder.plot()
+            # save plot if needed
+            if path_to_savefig:
+                fig = ax.get_figure()
+                fig.savefig(path_to_savefig)
+            # reset weights and optimizer state
+            if epochs_warmup != 0:
+                self.models[name].module.load_state_dict(state_cacher.retrieve("model"))
+                self.optims[name].load_state_dict(state_cacher.retrieve("optimizer"))
+                self.models[name].to(model_device)
+            else:
+                lr_finder.reset()
+
+            return optim_lr
+
+        assert mode == 'brute_force'
+        if num_epochs < 3:
+            raise ValueError("Number of epochs to find an optimal learning rate less than 3. It's pointless")
+        acc_store = dict()
+        state_cacher = StateCacher(in_memory=True, cache_dir=None)
+        state_cacher.store("model", self.models[name].module.cpu().state_dict())
+        # save on cpu, run on gpu
+        self.models[name].to(model_device)
+        state_cacher.store("optimizer", self.optims[name].state_dict())
+        range_lr = np.linspace(min_lr, max_lr, num_iter)
+        best_acc = 0.0
+        EPS = 0.017
+
+        for lr in range_lr:
+            print('Training with next lr: {}'.format(lr))
+            for param_group in self.optims[name].param_groups:
+                param_group["lr"] = round(lr,6)
+
+            top1 = round(self.run(max_epoch=num_epochs, lr_finder=True, start_eval=0, eval_freq=1,), 4)
+            acc_store[lr] = top1
+
+            self.models[name].module.load_state_dict(state_cacher.retrieve("model"))
+            self.optims[name].load_state_dict(state_cacher.retrieve("optimizer"))
+            self.models[name].to(model_device)
+            set_random_seed(seed)
+
+            cur_acc = acc_store[lr]
+            # break if the results got worse with epsilon confidence
+            if (best_acc - cur_acc) >= EPS:
```

Comment:
I don't think such comparisons of absolute values are always correct: may be we should use a relative improvement here. 
It's just a note for the future discussion, there is no need to change the exit criterion right now.